src/blocks/networkmanager.rs

```diff
@@ -28,125 +32,371 @@ enum NetworkState {
+
+        let device_type: Variant<u32> = m.get1().block_error("networkmanager", "Failed to read device type")?;
+        Ok(DeviceType::from(device_type.0))
+    }
+
+    fn active_access_point(&self, c: &Connection) -> Result<NMAccessPoint> {
+        let m = ConnectionManager::get(c, self.path.clone(), "org.freedesktop.NetworkManager.Device.Wireless", "ActiveAccessPoint")
+            .block_error("networkmanager", "Failed to retrieve device active access point")?;
+
+        let active_ap: Variant<Path> = m.get1().block_error("networkmanager", "Failed to read active access point")?;
+        Ok(NMAccessPoint { path: active_ap.0 })
+    }
+}
+
+#[derive(Clone)]
+struct NMAccessPoint<'a> {
+    path: Path<'a>,
+}
+
+impl<'a> NMAccessPoint<'a> {
+    fn ssid(&self, c: &Connection) -> Result<String> {
+        let m = ConnectionManager::get(c, self.path.clone(), "org.freedesktop.NetworkManager.AccessPoint", "Ssid").block_error("networkmanager", "Failed to retrieve SSID")?;
+
+        let ssid: Variant<Array<u8, Iter>> = m.get1().block_error("networkmanager", "Failed to read ssid")?;
+        Ok(std::str::from_utf8(&ssid.0.into_iter().collect::<Vec<u8>>())
+            .block_error("networkmanager", "Failed to parse ssid")?
+            .to_string())
+    }
+}
+
+#[derive(Clone)]
+struct NMIp4Config<'a> {
+    path: Path<'a>,
+}
+
+impl<'a> NMIp4Config<'a> {
+    fn addresses(&self, c: &Connection) -> Result<Vec<Ipv4Address>> {
+        let m = ConnectionManager::get(c, self.path.clone(), "org.freedesktop.NetworkManager.IP4Config", "Addresses").block_error("networkmanager", "Failed to retrieve addresses")?;
+
+        let addresses: Variant<Array<Array<u32, Iter>, Iter>> = m.get1().block_error("networkmanager", "Failed to read addresses")?;
+        Ok(addresses.0.into_iter().map(|addr| Ipv4Address::from(addr)).collect())
     }
 }
 
 pub struct NetworkManager {
     id: String,
-    output: TextWidget,
+    indicator: ButtonWidget,
+    output: Vec<ButtonWidget>,
     dbus_conn: Connection,
     manager: ConnectionManager,
-    show_type: bool,
+    config: Config,
+    on_click: Option<String>,
+    only_primary_connection: bool,
+    unknown_device_icon: bool,
+    ip: bool,
+    ssid: bool,
+    max_ssid_width: usize,
 }
 
 #[derive(Deserialize, Debug, Default, Clone)]
 #[serde(deny_unknown_fields)]
 pub struct NetworkManagerConfig {
-    /// Whether to show the connection type or not.
-    #[serde(default = "NetworkManagerConfig::default_show_type")]
-    pub show_type: bool,
+    #[serde(default = "NetworkManagerConfig::default_on_click")]
+    pub on_click: Option<String>,
+
+    /// Whether to only show the primary connection, or all active connections.
+    #[serde(default = "NetworkManagerConfig::default_only_primary_connection")]
+    pub only_primary_connection: bool,
```

Comment:
I would suggest this be simplified to `primary_only`.

Reply:
Noted, I'll update it.